File: kmp-nativecoroutines-core/src/appleMain/kotlin/com/rickclephas/kmp/nativecoroutines/NativeFlow.kt
package com.rickclephas.kmp.nativecoroutines

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import platform.Foundation.NSError
import kotlin.native.concurrent.freeze

/**
 * A function that collects a [Flow] via callbacks.
 *
 * The function takes an `onItem` and `onComplete` callback
 * and returns a cancellable that can be used to cancel the collection.
 */
typealias NativeFlow<T> = (onItem: NativeCallback<T>, onComplete: NativeCallback<NSError?>) -> NativeCancellable

/**
 * Creates a [NativeFlow] for this [Flow].
 *
 * @param scope the [CoroutineScope] to use for the collection, or `null` to use the [defaultCoroutineScope].
 * @receiver the [Flow] to collect.
 * @see Flow.collect
 */
fun <T> Flow<T>.asNativeFlow(scope: CoroutineScope? = null): NativeFlow<T> {
    val coroutineScope = scope ?: defaultCoroutineScope
    return (collect@{ onItem: NativeCallback<T>, onComplete: NativeCallback<NSError?> ->
        val job = coroutineScope.launch {
            try {
                collect { onItem(it.freeze()) }
                onComplete(null)
            } catch (e: Exception) {
                onComplete(e.asNSError())
            }
        }
        return@collect job.asNativeCancellable()
    }).freeze()
}

File: KMPNativeCoroutinesCore/NativeFlow.swift
//
//  NativeFlow.swift
//  KMPNativeCoroutinesCore
//
//  Created by Rick Clephas on 06/06/2021.
//

/// A function that collects a Kotlin coroutines Flow via callbacks.
///
/// The function takes an `onItem`, `onComplete` and `onCancelled` callback
/// and returns a cancellable that can be used to cancel the collection.
public typealias NativeFlow<Output, Failure: Error, Unit> = (
    _ onItem: @escaping NativeCallback2<Output, () -> Unit, Unit>,
    _ onComplete: @escaping NativeCallback<Failure?, Unit>,
    _ onCancelled: @escaping NativeCallback<Failure, Unit>
) -> NativeCancellable<Unit>

@available(*, deprecated, message: "Internal API used for Swift export source compatibility")
@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
public func nativeFlow<Sequence: AsyncSequence>(for asyncSequence: Sequence) -> NativeFlow<Sequence.Element, Error, Void> {
    { onItem, onComplete, onCancelled in
        let task = Task {
            do {
                for try await element in asyncSequence {
                    await withUnsafeContinuation { continuation in
                        onItem(element, continuation.resume, ())
                    }
                }
                onComplete(nil, ())
            } catch {
                if error is CancellationError {
                    onCancelled(error, ())
                } else {
                    onComplete(error, ())
                }
            }
        }
        return { task.cancel() }
    }
}
